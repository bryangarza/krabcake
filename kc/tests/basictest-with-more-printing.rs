use test_dependencies::VgKrabcakeClientRequest;
use test_dependencies::kc_borrow_mut;

pub fn main() {
    println!("Hello world (from `sb_rs_port/main.rs`)!");
    println!(
        "BorrowMut is {:x}",
        VgKrabcakeClientRequest::BorrowMut as u32
    );

    let mut val: u8 = 101;
    println!("about to &mut into x");
    let x = kc_borrow_mut!(val); // x = &mut val;
    println!("borrowed &mut val into x");
    let x_alias = x as *mut u8;
    println!("made alias of x in x_alias");
    println!("about to &mut into y");
    let y = kc_borrow_mut!(*x);
    println!("borrowed &mut val into y");

    println!("about to mutate val via y");
    *y = 105;
    println!("finished mutate val via y");

    println!("about to mutate val via x_alias");
    unsafe {
        *x_alias = 103;
    }
    println!("finished mutate val via x_alias");

    println!("about to read val via y");
    let end = *y;

    // Note: I didn't see a load against `y` above without a use
    // of `end` here. It would be nice to avoid requiring that,
    // but maybe such is life in release mode. Look into it.
    println!("Goodbye world, end: {}!", end);
}
