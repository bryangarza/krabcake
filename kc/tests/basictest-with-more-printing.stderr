lib.rs: normalizing output...
Hello world (from `rs_hello/src/lib.rs`)!
kc_main.c: dispatching code 4b430000
lib.rs: handle client request BORROW_MUT. borrowing_addr: 0x10000000
kc_main.c: dispatching code 4b430000
lib.rs: handle client request BORROW_MUT. borrowing_addr: 0x10000000
tag search seeking Unique(Counter(Counter(2))) and saw Unique(Counter(Counter(2)))
found tag in stack for address 0x10000000 when accessing via pointer with tag Counter(Counter(2)); stack len before: 2 after: 2
tag search seeking Unique(Counter(Counter(1))) and saw Unique(Counter(Counter(2)))
tag search seeking Unique(Counter(Counter(1))) and saw Unique(Counter(Counter(1)))
found tag in stack for address 0x10000000 when accessing via pointer with tag Counter(Counter(1)); stack len before: 2 after: 1
tag search seeking Unique(Counter(Counter(2))) and saw Unique(Counter(Counter(1)))
stack.tags has changed 1 -> 0, it's now: []
ALERT could not find tag in stack for address 0x10000000 even though we are accessing it via pointer with tag Counter(Counter(2))
